botocore/credentials.py

```diff
@@ -791,6 +820,193 @@ def _assume_role_base_kwargs(self, config):
+    return chosen[0] if chosen else None
+
+
+class AssumeRoleWithSamlProvider(AssumeRoleProvider):
+    METHOD = 'assume-role-with-saml'
+    DISTINCTION_VAR = 'saml_endpoint'
+
+    def __init__(self, load_config, client_creator, cache, profile_name,
+                 role_selector=_role_selector,
+                 password_prompter=getpass.getpass,
+                 username_prompter=raw_input,
+                 authenticators=None):
+        """
+        :type authenticators: list
+        :param authenticators: A list of authenticators, which are instances
+            with an is_suitable() method and an authenticate() method.
+            You can use it to add your own implementation for 3rd party IdP.
+        """
+        super(AssumeRoleWithSamlProvider, self).__init__(
+            load_config, client_creator, cache, profile_name)
+        self.username_prompter = username_prompter
+        self.password_prompter = password_prompter
+        self.role_selector = role_selector
+        self.authenticators = authenticators or [
+            SamlAdfsFormsBasedAuthenticator(),
+            SamlGenericFormsBasedAuthenticator()]
+
+    def _create_cache_key(self):
+        role_arn = self._get_role_config_values().get('role_arn')
+        cache_key = "%s--%s" % (self._profile_name, role_arn)
+        return cache_key.replace(':', '_').replace('/', '-')
+
+    def _get_role_config_values(self):
+        return self._loaded_config.get('profiles', {}).get(
+            self._profile_name, {})
+
+    def _create_creds_from_response(self, response):
+        config = self._get_role_config_values()
+        if config.get('saml_authentication_type') in ['form']:
+            # If some parameter(s) would require prompting the user for input,
+            # we use a different refresh_func.
+            # We can explore an option in the future to support reprompting for
+            # input, but for now we just error out when the temp creds expire.
+            refresh_func = lambda: exception_raiser(RefreshUnsupportedError)
+        else:
+            refresh_func = lambda: credential_normalizer(
+                self._create_client().assume_role_with_saml(
+                    **self._assume_role_base_kwargs(config))['Credentials'])
+        return RefreshableCredentials(
+            access_key=response['Credentials']['AccessKeyId'],
+            secret_key=response['Credentials']['SecretAccessKey'],
+            token=response['Credentials']['SessionToken'],
+            method=self.METHOD,
+            expiry_time=_parse_if_needed(
+                response['Credentials']['Expiration']),
+            refresh_using=refresh_func)
+
+    def _create_client(self):
+        # sts.assume_role_with_saml() requires no access keys,
+        # but we still need a pair of dummy values here to break recursion.
+        return self._client_creator(
+            'sts', aws_access_key_id='dummy', aws_secret_access_key='dummy')
```

Comment:
That's unfortunate. We should look into fixing that eventually, but not needed now.